agent/agent.go

```diff
@@ -0,0 +1,235 @@
+	// instructions for how to manage the client's balance. It should be
+	// displayed to the client. (Optional)
+	PoolMessageCallback func(string)
+
+	// NumHosts is the minimum number of vipnode hosts the client should
+	// maintain connections with. (Optional)
+	NumHosts int
+
+	// Payout is the address to register to associate pool credits towards.
+	// (Optional)
+	Payout string
+
+	mu       sync.Mutex
+	started  bool
+	stopCh   chan struct{}
+	waitCh   chan error
+	nodeInfo ethnode.UserAgent // cached during Start
+}
+
+// Start registers the node on the given pool and starts sending peer updates
+// every store.KeepaliveInterval. It returns after
+// successfully registering with the pool.
+func (a *Agent) Start(p pool.Pool) error {
+	a.mu.Lock()
+	if a.started {
+		a.mu.Unlock()
+		return errors.New("pool already started")
+	}
+	if a.stopCh == nil {
+		a.stopCh = make(chan struct{})
+		a.waitCh = make(chan error, 1)
+	}
+	a.mu.Unlock()
+
+	startCtx, cancel := context.WithTimeout(context.Background(), startTimeout)
+	defer cancel()
+
+	enode, err := a.EthNode.Enode(startCtx)
+	if err != nil {
+		return err
+	}
+	logger.Printf("Connected to local node: %s", enode)
+
+	version := a.Version
+	if version == "" {
+		version = "dev"
+	}
+
+	connectReq := pool.ConnectRequest{
+		Payout:         a.Payout,
+		NodeURI:        a.NodeURI,
+		VipnodeVersion: version,
+		NodeInfo:       a.EthNode.UserAgent(),
+	}
+	a.nodeInfo = connectReq.NodeInfo
+	resp, err := p.Connect(startCtx, connectReq)
+	if err != nil {
+		return err
+	}
+	logger.Printf("Registered on pool: Version %s", resp.PoolVersion)
+
+	if resp.Message != "" && a.PoolMessageCallback != nil {
+		a.PoolMessageCallback(resp.Message)
+	}
+
+	if err := a.updatePeers(startCtx, p); err != nil {
+		return err
+	}
+
+	go func() {
+		a.waitCh <- a.serveUpdates(p)
+	}()
+	return nil
+}
+
+// Whitelist a peer for this node.
+func (a *Agent) Whitelist(ctx context.Context, nodeID string) error {
+	logger.Printf("Received whitelist request: %s", nodeID)
+	return a.EthNode.AddTrustedPeer(ctx, nodeID)
+}
+
+// Stop shuts down all the active connections cleanly.
+func (a *Agent) Stop() {
+	a.stopCh <- struct{}{}
+}
+
+// Wait blocks until the agent is stopped. It returns any errors that occur
+// during stopping.
+func (a *Agent) Wait() error {
+	return <-a.waitCh
+}
+
+func (a *Agent) serveUpdates(p pool.Pool) error {
+	ticker := time.Tick(store.KeepaliveInterval)
+	for {
+		select {
+		case <-ticker:
+			// FIXME: Sametimes a.updatePeers could take a while, does it make
+			// sense to keep ticking every interval regardless? Or should we
+			// make sure there is an offset between calls? What if there's
+			// overlap, do we want a mutex?
+			if err := a.updatePeers(context.Background(), p); err != nil {
+				// FIXME: Does it make sense to continue updating for certain
+				// errors? Eg if no hosts are found, we could keep sending
+				// updates until we find some.
```

Comment:
As long as the error gets propagated all the way up to the agent stopping, I think this is acceptable behavior, since I think it's sane to assume the agent would be run under some sort of supervisor that can restart it, and these errors should be pretty uncommon right?

Reply:
I actually ended up changing updatePeers to not return an error if no hosts are found, only if the RPC transport failed. I think this is the way to go, I'll remove the comment. Thanks!